stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/repository/domain/channel/internal/DatabaseChannelRepository.kt

```diff
@@ -16,52 +16,67 @@
 
 package io.getstream.chat.android.offline.repository.domain.channel.internal
 
-import io.getstream.chat.android.client.extensions.internal.lastMessage
+import android.util.LruCache
 import io.getstream.chat.android.client.persistance.repository.ChannelRepository
+import io.getstream.chat.android.core.utils.date.maxOf
+import io.getstream.chat.android.core.utils.date.minOf
 import io.getstream.chat.android.models.Channel
 import io.getstream.chat.android.models.Member
 import io.getstream.chat.android.models.Message
 import io.getstream.chat.android.models.User
-import io.getstream.chat.android.offline.repository.domain.channel.lastMessageInfo
-import io.getstream.chat.android.offline.repository.domain.channel.member.internal.toEntity
-import io.getstream.chat.android.offline.repository.domain.channel.member.internal.toModel
+import io.getstream.chat.android.offline.extensions.awaitWithMutex
+import io.getstream.chat.android.offline.extensions.launchWithMutex
 import io.getstream.log.taggedLogger
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.sync.Mutex
 import java.util.Date
 
 /**
  * Repository to read and write [Channel] data.
  */
 @SuppressWarnings("TooManyFunctions")
 internal class DatabaseChannelRepository(
+    private val scope: CoroutineScope,
     private val channelDao: ChannelDao,
     private val getUser: suspend (userId: String) -> User,
     private val getMessage: suspend (messageId: String) -> Message?,
     cacheSize: Int = 1000,
 ) : ChannelRepository {
 
     private val logger by taggedLogger("Chat:ChannelRepository")
+    private val channelCache = LruCache<String, Channel>(cacheSize)
+    private val dbMutex = Mutex()
 
-    /**
-     * Inserts a [Channel]
-     *
-     * @param channel [Channel]
-     */
-    override suspend fun upsertChannel(channel: Channel) {
-        val entity = channel.convertToEntity()
-        logger.v { "[upsertChannel] entity: ${entity.lastMessageInfo()}" }
-        channelDao.insert(entity)
+    override suspend fun insertChannel(channel: Channel) {
+        insertChannels(listOf(channel))
     }
 
     /**
      * Inserts many [Channel]s.
      *
      * @param channels collection of [Channel]
      */
-    override suspend fun upsertChannels(channels: Collection<Channel>) {
+    override suspend fun insertChannels(channels: Collection<Channel>) {
         if (channels.isEmpty()) return
-        val entities = channels.map { channel -> channel.convertToEntity() }
-        logger.v { "[upsertChannels] entities.size: ${entities.size}" }
-        channelDao.insertMany(entities)
+        val updatedChannels = channels
+            .map { channelCache[it.cid]?.let { cachedChannel -> it.combine(cachedChannel) } ?: it }
+        val channelToInsert = updatedChannels
+            .filter { channelCache[it.cid] != it }
+            .map { it.toEntity() }
+        cacheChannel(updatedChannels)
```

Comment:
I think there could be a case where the number of channels is bigger than 1000 - in that case not all of them will be cached. And this function (like some others in the PR) will return immediately but the cache will not be a complete representation of the state. Can this cause some consistency issues? 🤔  If so - can we detect a case where it doesn't fit the cache and we should fire the DB calls synchronously?...

Reply:
In that case, a "double insertion" of the same channels (Channel number 1001 and the following ones).
Not a real case for our customers, it shouldn't be so big and as far as we only use it as "offline support", it will only have some data of the channel that is not up-to-date for a moment but updated whenever a new event arrives on that channel